src/test/java/net/snowflake/client/jdbc/PreparedStatementIT.java

```diff
@@ -0,0 +1,1721 @@
+    }
+  }
+
+  @Test
+  @ConditionalIgnore(condition = RunningOnTravisCI.class)
+  public void testStageString() throws SQLException
+  {
+    connection = getConnection();
+    int[] thresholds = {0, 6}; // disabled, enabled
+    String[] rows = {null, "", "\"", ",", "\n", "\r\n", "\"\"", "null", "\\\n", "\",", "\\\",\\\""};
+
+    for (int threshold : thresholds)
+    {
+      connection.createStatement().execute("DELETE FROM TEST_PREPST WHERE 1=1"); // clear table
+      connection.createStatement().execute(
+          String.format("ALTER SESSION SET CLIENT_STAGE_ARRAY_BINDING_THRESHOLD = %d", threshold));
+      prepStatement = connection.prepareStatement(insertSQL);
+      for (int i = 0; i < rows.length; i++)
+      {
+        bindOneParamSet(prepStatement, i, 0.0, 0.0f, rows[i], 0, (short) 0);
+        prepStatement.addBatch();
+      }
+      prepStatement.executeBatch();
+
+      resultSet = connection.createStatement().executeQuery("SELECT colC FROM TEST_PREPST ORDER BY id ASC");
+      String errorMessage = "Strings should match (" + (threshold > 0 ? "stage" : "non-stage") + ")";
+
+      for (int i = 0; i < rows.length; i++)
+      {
+        resultSet.next();
+        assertEquals(errorMessage, rows[i], resultSet.getString(1));
+      }
+    }
+  }
+
+  @Test
+  @ConditionalIgnore(condition = RunningOnTravisCI.class)
+  public void testIncorrectTypes() throws SQLException
+  {
+    connection = getConnection();
+    int[] thresholds = {0, 6}; // disabled, enabled
+
+    for (int threshold : thresholds)
+    {
+      connection.createStatement().execute("DELETE FROM TEST_PREPST WHERE 1=1"); // clear table
+      connection.createStatement().execute(
+          String.format("ALTER SESSION SET CLIENT_STAGE_ARRAY_BINDING_THRESHOLD = %d", threshold));
+      prepStatement = connection.prepareStatement(insertSQL);
+
+      prepStatement.setString(1, "notAnInt"); // should cause error
+      prepStatement.setDouble(2, 0.0);
+      prepStatement.setFloat(3, 0.0f);
+      prepStatement.setString(4, "");
+      prepStatement.setLong(5, 0);
+      prepStatement.setShort(6, (short) 0);
+      prepStatement.addBatch();
+
+      try
+      {
+        prepStatement.executeBatch();
+        fail("An exception should have been thrown");
+      }
+      catch (SQLException ex)
+      {
+      }
+    }
+  }
+
+  @Test
+  @ConditionalIgnore(condition = RunningOnTravisCI.class)
+  public void testStageBatchTimestamps() throws SQLException
+  {
+    connection = getConnection();
+    Timestamp tsEpoch = new Timestamp(0L);
+    Timestamp tsEpochMinusOneSec = new Timestamp(-1000L); // negative epoch no fraction of seconds
+    Timestamp tsPast = new Timestamp(-2208988800100L); // very large negative epoch
+    Timestamp tsFuture = new Timestamp(32503680000000L); // very large positive epoch
+    Timestamp tsNow = new Timestamp(System.currentTimeMillis());
+    Timestamp tsArbitrary = new Timestamp(862056000000L);
```

Comment:
Be create to add more variation of tests here. Currently this covers:
- epoch time
- negative epoch time without fraction of seconds
- large negative epoch time with fraction of seconds
- large positive epoch time without fraction of seconds
- current time
- general time
- null

Reply:
I will add more in a separate commit.